src/inbox/chan_ptr.rs

```diff
@@ -0,0 +1,385 @@
+            .trim_start_matches("::");
+        f.debug_struct(&format!("ChanPtr<{}, {}>", act, rc))
+            .field("rx_count", &self.inner.receiver_count.load(SeqCst))
+            .field("tx_count", &self.inner.sender_count.load(SeqCst))
+            .finish()
+    }
+}
+
+/// The reference count of a strong address. Strong addresses will prevent the actor from being
+/// dropped as long as they live. Read the docs of [`Address`](crate::Address) to find
+/// out more.
+#[derive(Debug)]
+pub struct TxStrong(());
+
+impl TxStrong {
+    /// Attempt to construct a new `TxStrong` pointing to the given `inner` if there are existing
+    /// strong references to `inner`. This will return `None` if there were 0 strong references to
+    /// the inner.
+    fn try_new<A>(inner: &Chan<A>) -> Option<TxStrong> {
+        // All code taken from Weak::upgrade in std
+        use std::sync::atomic::Ordering::*;
+
+        // Relaxed load because any write of 0 that we can observe leaves the field in a permanently
+        // zero state (so a "stale" read of 0 is fine), and any other value is confirmed via the
+        // CAS below.
+        let mut n = inner.sender_count.load(Relaxed);
+
+        loop {
+            if n == 0 {
+                return None;
+            }
+
+            // Relaxed is fine for the failure case because we don't have any expectations about the new state.
+            // Acquire is necessary for the success case to synchronise with `Arc::new_cyclic`, when the inner
+            // value can be initialized after `Weak` references have already been created. In that case, we
+            // expect to observe the fully initialized value.
+            match inner
+                .sender_count
+                .compare_exchange_weak(n, n + 1, Acquire, Relaxed)
+            {
+                Ok(_) => return Some(TxStrong(())), // 0-case checked above
+                Err(old) => n = old,
+            }
+        }
+    }
+}
+
+/// The reference count of a weak address. Weak addresses will bit prevent the actor from being
+/// dropped. Read the docs of [`Address`](crate::Address) to find out more.
+#[derive(Debug)]
+pub struct TxWeak(());
+
+/// A reference counter that can be dynamically either strong or weak.
+#[derive(Debug)]
+pub enum TxEither {
+    /// A strong reference counter.
+    Strong(TxStrong),
+    /// A weak reference counter.
+    Weak(TxWeak),
+}
+
+/// A reference counter for the receiving end of the channel or in actor terminology, the mailbox
+/// of an actor.
+pub struct Rx(());
+
+/// Defines a reference counting policy for a channel pointer.
+///
+/// # Implementation note
+///
+/// This trait only exists because we cannot specialise `Drop` impls in Rust, otherwise, we could
+/// write:
+///
+/// - `impl Drop for ChanPtr<A, TxStrong> { }`
+/// - `impl Drop for ChanPtr<A, TxEither> { }`
+/// - `impl Drop for ChanPtr<A, Rx> { }`
+///
+/// and call the appropriate functions on `Chan`.
```

Comment:
```suggestion
/// and call the appropriate functions on `Chan`.
///
/// It can also be used to generalise over the reference counter type in extension trait implementations on `Address`. For example:
/// ```
/// # use xtra::{Actor, Address, RefCounter};
/// trait MyExtension {
///    fn say_hi(&self) {
///        println!("Hi!"); // Presumably, you would do something useful here
///    }
/// }
///
/// impl<A, Rc> MyExtension for Address<A, Rc>
///    where A: Actor,
///         Rc: RefCounter
/// {}
```

Reply:
I am not sure if this is formatted correctly or compiles :sweat_smile:

Reply:
Cool, thanks! I'll fix it up :)